docs/src/main/asciidoc/credentials-provider.adoc

```diff
@@ -0,0 +1,194 @@
+if credentials are generated (e.g. `Vault Dynamic DB Credentials`) or if a custom credentials provider is required.
+
+This guide will show how to use the `Credentials Provider` provided in the `vault` extension,
+then we will look at implementing a custom `Credentials Provider`, and finally we will talk about additional
+considerations regarding implementing a `Credentials Provider` in a new extension.
+
+include::./status-include.adoc[]
+
+== Prerequisites
+
+To complete this guide, you need:
+
+* roughly 20 minutes
+* an IDE
+* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
+* Apache Maven {maven-version}
+* Docker installed
+
+== Vault Credentials Provider
+
+To configure a `Vault Credentials Provider` you need to provide the following property(ies):
+
+[source, properties]
+----
+quarkus.vault.credentials-provider.<name>.<property>=<value>
+----
+
+The `<name>` will be used in the consumer to refer to this provider. The `<property>` and `<value>` fields are specific to the `Vault Credentials Provider`. For complete details, please refer to the {vault-datasource-guide}.
+
+For instance:
+
+[source, properties]
+----
+quarkus.vault.credentials-provider.mydatabase.kv-path=myapps/vault-quickstart/db
+----
+
+Once defined, the `mydatabase` provider can be used in any extension that supports the `Credentials Provider` interface. For instance in `agroal`:
+
+[source, properties]
+----
+# configure your datasource
+quarkus.datasource.db-kind = postgresql
+quarkus.datasource.username = sarah
+quarkus.datasource.credentials-provider = mydatabase
+quarkus.datasource.jdbc.url = jdbc:postgresql://localhost:5432/mydatabase
+----
+
+Note that `quarkus.datasource.username` is the original `agroal` property, whereas the `password` property
+is not included because the value will come from the `mydatabase` credentials provider we just defined.
+An alternative is to define both username and password in Vault and drop the `quarkus.datasource.username`
+property from configuration. All consuming extensions do support the ability to fetch both the username
+and password from the provider, or just the password.
+
+== Custom Credentials Provider
+
+Implementing a custom credentials provider is the only option when a vault product is not yet supported in Quarkus, or if credentials need to be retrieved from a custom store.
+
+The only interface to implement is:
+
+[source, java]
+----
+public interface CredentialsProvider {
+
+    String USER_PROPERTY_NAME = "user";
+    String PASSWORD_PROPERTY_NAME = "password";
+
+    Map<String, String> getCredentials(String credentialsProviderName);
+
+}
+----
+
+`USER_PROPERTY_NAME` and `PASSWORD_PROPERTY_NAME` are standard properties that should be recognized by any consuming extension that support username/password based authentication.
+
+Here is a simple implementation example:
+
+[source, java]
+----
+@ApplicationScoped
+@Unremovable
```

Comment:
I think these need a little explaining as it's not mentioned in the docs below or above that the implementation of `CredentialsProvider` needs to be a CDI bean

Reply:
I have added:

> It is required that implementations be valid `ApplicationScoped` CDI beans.
> 
> Here is a simple example:
> ...